File: tests/st/dyn_shape_dev/test_ops_expand_dims.py
# Copyright 2023 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================

import pytest
import numpy as np
import mindspore as ms
from mindspore import ops, Tensor


@ms.jit
def expand_dims_forward_func(x, axis):
    return ops.auto_generate.expand_dims(x, axis)


@ms.jit
def expand_dims_backward_func(x, axis):
    return ops.grad(expand_dims_forward_func, (0,))(x, axis)


@pytest.mark.level0
@pytest.mark.env_onecard
@pytest.mark.platform_x86_cpu
@pytest.mark.platform_x86_gpu_training
@pytest.mark.platform_arm_ascend_training
def test_expand_dims_forward():
    """
    Feature: Ops.
    Description: test op expand_dims.
    Expectation: expect correct result.
    """
    x = Tensor(np.array([[2, 2], [2, 2]]), ms.float32)
    axis = 0
    output = expand_dims_forward_func(x, axis)
    expect = [[[2., 2.], [2., 2.]]]
    assert np.allclose(output.asnumpy(), expect, rtol=0.001)


@pytest.mark.level0
@pytest.mark.env_onecard
@pytest.mark.platform_x86_cpu
@pytest.mark.platform_x86_gpu_training
@pytest.mark.platform_arm_ascend_training
def test_expand_dims_backward():
    """
    Feature: Auto grad.
    Description: test auto grad of op expand_dims.
    Expectation: expect correct result.
    """
    x = Tensor(np.array([[2, 2], [2, 2]]), ms.float32)
    axis = 0
    output = expand_dims_backward_func(x, axis)
    expect = [[1., 1.], [1., 1.]]
    assert np.allclose(output.asnumpy(), expect, rtol=0.001)


@pytest.mark.level0
@pytest.mark.env_onecard
@pytest.mark.platform_x86_cpu
@pytest.mark.platform_x86_gpu_training
@pytest.mark.platform_arm_ascend_training
def test_expand_dims_vmap():
    """
    Feature: test vmap function.
    Description: test expand_dims op vmap.
    Expectation: expect correct result.
    """
    x = Tensor(np.array([[[[2, 2], [2, 2]]]]), ms.float32)
    axis = 0
    nest_vmap = ops.vmap(ops.vmap(expand_dims_forward_func, in_axes=(0, None)), in_axes=(0, None))
    output = nest_vmap(x, axis)
    expect = [[[[[2., 2.], [2., 2.]]]]]
    assert np.allclose(output.asnumpy(), expect, rtol=1e-4, atol=1e-4)

File: tests/ut/cpp/ops/test_ops_expand_dims.cc
/**
 * Copyright 2023 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <vector>
#include <memory>
#include "common/common_test.h"
#include "ir/dtype/type.h"
#include "abstract/dshape.h"
#include "utils/tensor_construct_utils.h"
#include "ir/primitive.h"
#include "abstract/abstract_value.h"
#include "include/backend/optimizer/helper.h"
#include "ops/test_ops.h"
#include "ops/ops_func_impl/expand_dims.h"
#include "ops/test_value_utils.h"

namespace mindspore {
namespace ops {
struct ExpandDimsShapeParams {
  ShapeVector input_x_shape;
  TypePtr input_x_type;
  ValuePtr axis;
  ShapeVector output_shape;
  TypePtr output_type;
};

class TestExpandDims : public TestOps, public testing::WithParamInterface<ExpandDimsShapeParams> {};

TEST_P(TestExpandDims, dyn_shape) {
  const auto &param = GetParam();
  auto input_x = std::make_shared<abstract::AbstractTensor>(param.input_x_type, param.input_x_shape);
  auto expect_shape = std::make_shared<abstract::Shape>(param.output_shape);
  auto expect_dtype = std::make_shared<TensorType>(param.output_type);

  ExpandDimsFuncImpl expand_dims_func_impl;
  auto prim = std::make_shared<Primitive>("ExpandDims");

  auto out_dtype = expand_dims_func_impl.InferType(prim, {input_x, param.axis->ToAbstract()});
  ASSERT_TRUE(*out_dtype == *expect_dtype);
  auto out_shape = expand_dims_func_impl.InferShape(prim, {input_x, param.axis->ToAbstract()});
  ASSERT_TRUE(*out_shape == *expect_shape);
}

INSTANTIATE_TEST_CASE_P(
  TestExpandDims, TestExpandDims,
  testing::Values(ExpandDimsShapeParams{{3, 4, 5}, kFloat32, CreateScalar<int64_t>(2), {3, 4, 1, 5}, kFloat32},
                  ExpandDimsShapeParams{{3, 4, 5}, kInt64, CreateScalar<int64_t>(0), {1, 3, 4, 5}, kInt64},
                  ExpandDimsShapeParams{{3, 4, 5}, kInt64, CreateScalar<int64_t>(-3), {3, 1, 4, 5}, kInt64},
                  ExpandDimsShapeParams{{2, 3, 4, 5}, kInt32, CreateScalar<int64_t>(-2), {2, 3, 4, 1, 5}, kInt32},
                  ExpandDimsShapeParams{{-1, -1, -1}, kUInt64, CreateScalar<int64_t>(2), {-1, -1, 1, -1}, kUInt64},
                  ExpandDimsShapeParams{{-1, -1, -1}, kFloat32, CreateScalar(kValueAny), {-1, -1, -1, -1}, kFloat32},
                  ExpandDimsShapeParams{{-2}, kFloat32, CreateScalar<int64_t>(2), {-2}, kFloat32}));
}  // namespace ops
}  // namespace mindspore
